src/main/java/com/slack/kaldb/server/Kaldb.java

```diff
@@ -50,104 +47,109 @@ public Kaldb(Path configFilePath) throws IOException {
-
-      Server server = sb.build();
-
-      CompletableFuture<Void> serverFuture = server.start();
-      serverFuture.join();
-      LOG.info("Started indexer server on port: {}", serverPort);
-
-      // TODO: Instead of passing in the indexer, consider creating an interface or make indexer of
-      // subclass of this class?
-      indexer.start();
+      ArmeriaService armeriaService =
+          new ArmeriaService(serverPort, prometheusMeterRegistry, searcher, "armeriaIndexService");
+      services.add(armeriaService);
     }
 
     if (roles.contains(KaldbConfigs.NodeRole.QUERY)) {
-      ServerBuilder sb = Server.builder();
-
-      GrpcServiceBuilder searchBuilder =
-          GrpcService.builder()
-              .addService(new KaldbDistributedQueryService())
-              .enableUnframedRequests(true);
-      sb.service(searchBuilder.build());
-
+      KaldbDistributedQueryService searcher = new KaldbDistributedQueryService();
       final int serverPort = KaldbConfig.get().getQueryConfig().getServerPort();
-      addManagementEndpoints(sb, serverPort);
-
-      Server server = sb.build();
-
-      CompletableFuture<Void> serverFuture = server.start();
-      serverFuture.join();
-      LOG.info("Started query server on port: {}", serverPort);
+      ArmeriaService armeriaService =
+          new ArmeriaService(serverPort, prometheusMeterRegistry, searcher, "armeriaQueryService");
+      services.add(armeriaService);
     }
 
-    // TODO: On CTRL-C shut down the process cleanly. Ensure no write locks in indexer. Guava
-    // ServiceManager?
+    return services;
   }
 
-  private LoggingServiceBuilder getLoggingServiceBuilder() {
-    return LoggingService.builder()
-        // Not logging any successful response, say prom scraping /metrics every 30 seconds at INFO
-        .successfulResponseLogLevel(LogLevel.DEBUG)
-        .failureResponseLogLevel(LogLevel.ERROR)
-        // Remove all headers to be sure we aren't leaking any auth/cookie info
-        .requestHeadersSanitizer((ctx, headers) -> DefaultHttpHeaders.EMPTY_HEADERS);
+  public static ServiceManager.Listener getServiceManagerListener() {
+    return new ServiceManager.Listener() {
+      @Override
+      public void failure(Service service) {
+        LOG.error(
+            String.format(
+                "Service %s failed with cause %s",
+                service.getClass().toString(), service.failureCause().toString()));
+        LOG.error("Shutting down Kaldb due to failed service");
+        // shutdown if any services enters failure state
+        System.exit(1);
+      }
+    };
   }
 
-  private void setupSystemMetrics() {
+  public static void addShutdownHook(ServiceManager serviceManager) {
+    Runtime.getRuntime()
+        .addShutdownHook(
+            new Thread(
+                () -> {
+                  try {
+                    serviceManager.stopAsync().awaitStopped(30, TimeUnit.SECONDS);
+
+                    // then shut down log4j
```

Comment:
Please add a comment about log4j shutdown is done here so we can log all the server shutdown logs. That way, we can have the needed context.

Reply:
Done!